### PR TITLE
Rewards carousel navigation bar divider line handling

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Some UI tweaks to `RewardPledgeNavigationController`'s navigation bar to polish the feel of the carousel. We hide and show the divider line of the navigation bar as we pan the reward up and down. There are some edge cases that we needed to handle based on which feature flags are enabled.

**Note:** There is some additional finessing required when we enable the native pledge view, I have created a ticket to track this (it's specific to the navigation bar colour during the transition).

# 🤔 Why

Continuation of the native checkout rewards carousel work.

# 🛠 How

- Apply the same background colour as the `RewardsCollectionViewController`'s view to the `UINavigationBar` so that it appears as one view.
- Set an empty `UIImage()` as the `shadowImage` property of `UINavigationBar` to hide the divider line.
- Hide and show the divider line when necessary.

# 👀 See

Trello, screenshots, external resources?

| Native Checkout **ON**, Native Pledge View **OFF** (Divider appears, hides on pop) | Native Checkout **OFF**, Native Pledge View **OFF** (Divider appears normally) |
| --- | --- |
|  |  |

# ✅ Acceptance criteria

- [ ] Native Checkout ON, Native Pledge View OFF: Divider line appears as you pan the reward under the nav bar. It disappears when the reward is not being clipped by the nav bar. If you select a reward, the divider is shown on the deprecated pledge view. When you pop that view, the divider line returned to whichever state it had before the push.
- [ ] Native Checkout OFF, Native Pledge View OFF: No regression on current app behaviour.

**Note:** If both feature flags are enabled there is some additional work required as mentioned above which will be handled separately.

### CODE REVIEW

File: Kickstarter-iOS/Views/Cells/RewardCell.swift [hunk 1]
```diff
@@ -105,14 +112,24 @@ final class RewardCell: UICollectionViewCell, ValueCell {
...
     }
 
     animator.startAnimation()
   }
 }
 
+extension RewardCell: UIScrollViewDelegate {
+  func scrollViewDidScroll(_ scrollView: UIScrollView) {
+    let cardContainerViewY = self.rewardCardContainerView.frame.origin.y
+    let yOffset = scrollView.contentOffset.y - scrollViewTopInset - cardContainerViewY
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Cells/RewardCell.swift]: Are you able to get the `scrollViewTopInset` dynamically from ` self.scrollView.contentInset.top`?

File: Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift [hunk 2]
```diff
@@ -307,13 +324,31 @@ final class RewardsCollectionViewController: UICollectionViewController {
...
+  // MARK: - Actions
 
   @objc func closeButtonTapped() {
     self.navigationController?.dismiss(animated: true)
   }
 }
 
+// MARK: - Functions
+
+private func newShadowImage(with color: UIColor) -> UIImage? {
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift]: Could this be an extension on `UIImage`?

```
let color = UIColor.ksr_grey_200
UIImage.image(with: color)
```

Comment (justinswart) [reply Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift]: Could be... wasn't sure if we'd need it anywhere else?

File: Kickstarter-iOS/Views/Cells/RewardCell.swift [hunk 3]
```diff
@@ -105,14 +110,26 @@ final class RewardCell: UICollectionViewCell, ValueCell {
         transform = .identity
       }
 
-      _ = self
+      _ = self.rewardCardContainerView
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Cells/RewardCell.swift]: What's this for? 👀

Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/RewardCell.swift]: Previously we were transforming the whole cell which was causing a visual gap between the card and this divider line. Now we just transform the `rewardCardContainerView` to fix that.

Comment (ifbarrera) [reply Kickstarter-iOS/Views/Cells/RewardCell.swift]: Nice! Thanks for fixing 🙏

File: Kickstarter-iOS/Views/Cells/PillCell.swift [hunk 4]
```diff
@@ -34,6 +34,7 @@ final class PillCell: UICollectionViewCell, ValueCell {
 
     _ = self.contentView
       |> contentViewStyle
+      |> \.backgroundColor .~ .ksr_grey_200
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Cells/PillCell.swift]: Is this right? Shouldn't the background be green?

image

File: Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift [hunk 5]
```diff
@@ -68,6 +68,8 @@ final class RewardsCollectionViewController: UICollectionViewController {
   override func viewDidLoad() {
     super.viewDidLoad()
 
+    self.extendedLayoutIncludesOpaqueBars = true
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift]: Can we use the `keyPath` setters?

File: Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift [hunk 6]
```diff
@@ -87,23 +89,22 @@ final class RewardsCollectionViewController: UICollectionViewController {
...
-
   override func viewWillAppear(_ animated: Bool) {
     super.viewWillAppear(animated)
 
     self.viewModel.inputs.viewWillAppear()
   }
 
-  override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
-    super.viewWillTransition(to: size, with: coordinator)
+  override func viewDidLayoutSubviews() {
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift]: Hmm this shouldn't be new, it's in master. I think you need to pull in the latest from master.

Comment (justinswart) [reply Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift]: Yeah I merged, I think it just moved in the merge. lemme move it above `viewWillAppear`, should make it go away.

File: Kickstarter-iOS/Views/Cells/PillCell.swift (hunk 4 above)
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/PillCell.swift]: Apologies 🤦‍♂

**Pull request merged**